content/blog/learn-pulumi-inputs-outputs/index.md

```diff
@@ -0,0 +1,160 @@
+
+![Load balancer example](i_o-example.png)
+
+The diagram shows how this works. Assume that we create a virtual machine that has an IP address as its output.  Next, we'll create a load balancer that has a target and a port as inputs. We'll pass the virtual machine's IP address as the target because we want the load balancer to point at it, and then we'll set 80 as the input port. The resulting load balancer's outputs will then be the generated hostname and the port 80.
+
+The diagram shows the dependency. The load balancer knows that it depends on the virtual machine because the target points at the virtual machine's IP address. Be aware that when you're running `pulumi up`, you may encounter outputs. In this case, we're creating
+the virtual machine and load balancer, and if we examine details in the preview, we'll see the target is listed as an output of string. That's because, during the preview, its value isn't known. In this case, the port is known as 80 because it’s a concrete value, but it’s possible that the port could come from another resource's output, or it was computed. In that case, it might be shown as an output of type number.
+
+Let's look at some code. This worked example uses AWS resources and presents a slightly more complex situation. However, the way inputs and outputs are used are the same as the basic example.
+
+The first thing to notice is that the code is wrapped in an `async ()` function. The reason for this is that we create a Virtual Private Cloud (VPC) which defines a virtual network for our application.  In place of an IP address as a target, the load balancer uses the VPC's public subnets to communicate to the load balancer, aren’t known in advance. We also create a security group to control egress from the VPC.
+
+```typescript
+import * as aws from "@pulumi/aws";
+import * as awsx from "@pulumi/awsx";
+import * as pulumi from "@pulumi/pulumi";
+
+export = async () => {
+    const config = new pulumi.Config("aws");
+    const providerOpts = { provider: new aws.Provider("prov", { region: <aws.Region>config.require("region") }) };
+
+    // Create a vpc
+    const vpc = awsx.ec2.Vpc.getDefault();
+
+    // Create a security group to open ingress to our load balancer on port 80, and egress out of the VPC.
+    const sg = new awsx.ec2.SecurityGroup("web-sg", {
+        vpc,
+        egress: [{ protocol: "-1", fromPort: 0, toPort: 0, cidrBlocks: [ "0.0.0.0/0" ] }],
+    });
+```
+
+In this part of the code, the application load balancer is created, and port 80 is assigned to the load balancer's listener. Note that we create `publicIps` as an Output typed as a string array, which will hold the virtual machines' public IP addresses. We also call `await` to return the public subnets as an output, because we need that information before creating the virtual machines.
```

Comment:
If we have a load balancer, why do we need the VMs to have public IPs? Why do we even need more than one VM? 

I still think this example is significantly more complicated than it needs to be to illustrate these concepts; the inputs and outputs are getting lost in all the infrastructure swirling around here. (Sorry to be a pain about it, but I do believe it.)